File: ios/RocChat/Auth/APIClient.swift
import Foundation
import CommonCrypto
import CryptoKit
import Security

private final class PinnedSessionDelegate: NSObject, URLSessionDelegate {
    static let shared = PinnedSessionDelegate()

    private let pinnedLeafFingerprints: [String: String] = [
        "chat.mocipher.com": "9fe88e6203b5c859d5d5afad6b2efe52b3f01bb06ea1b140c43b1b77ebd89dbb",
        "rocchat-api.spoass.workers.dev": "fced5d78a8da9d40a74f053b91aed908a2a0ef097a8878b002409be9b35eead1",
    ]

    func urlSession(_ session: URLSession, didReceive challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
        guard challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust,
              let trust = challenge.protectionSpace.serverTrust else {
            completionHandler(.performDefaultHandling, nil)
            return
        }

        let host = challenge.protectionSpace.host.lowercased()
        guard let expectedFingerprint = pinnedLeafFingerprints[host] else {
            completionHandler(.performDefaultHandling, nil)
            return
        }

        var trustError: CFError?
        guard SecTrustEvaluateWithError(trust, &trustError) else {
            completionHandler(.cancelAuthenticationChallenge, nil)
            return
        }
        // SecTrustGetCertificateAtIndex was deprecated in iOS 15; use the chain copy instead.
        guard let chain = SecTrustCopyCertificateChain(trust) as? [SecCertificate],
              let certificate = chain.first else {
            completionHandler(.cancelAuthenticationChallenge, nil)
            return
        }

        let certificateData = SecCertificateCopyData(certificate) as Data
        let actualFingerprint = SHA256.hash(data: certificateData).map { String(format: "%02x", $0) }.joined()

        guard actualFingerprint == expectedFingerprint else {
            completionHandler(.cancelAuthenticationChallenge, nil)
            return
        }

        completionHandler(.useCredential, URLCredential(trust: trust))
    }
}

class APIClient {
    static let shared = APIClient()
    
    var sessionToken: String?
    var refreshToken: String?
    private let baseURL: String
    let session: URLSession
    
    init(baseURL: String = "https://chat.mocipher.com/api") {
        self.baseURL = baseURL
        let configuration = URLSessionConfiguration.ephemeral
        configuration.waitsForConnectivity = true
        self.session = URLSession(configuration: configuration, delegate: PinnedSessionDelegate.shared, delegateQueue: nil)
        // Migrate session_token from UserDefaults to SecureStorage if needed
        if let legacySession = UserDefaults.standard.string(forKey: "session_token"),
           SecureStorage.shared.get(forKey: "session_token") == nil {
            SecureStorage.shared.set(legacySession, forKey: "session_token")
            UserDefaults.standard.removeObject(forKey: "session_token")
        }
        self.sessionToken = SecureStorage.shared.get(forKey: "session_token")
        // Migrate refresh_token from UserDefaults to SecureStorage if needed
        if let legacyRefresh = UserDefaults.standard.string(forKey: "refresh_token"),
           SecureStorage.shared.get(forKey: "refresh_token") == nil {
            SecureStorage.shared.set(legacyRefresh, forKey: "refresh_token")
            UserDefaults.standard.removeObject(forKey: "refresh_token")
        }
        self.refreshToken = SecureStorage.shared.get(forKey: "refresh_token")
    }

    func webSocketTask(with url: URL) -> URLSessionWebSocketTask {
        session.webSocketTask(with: url)
    }
    
    struct LoginResult {
        let sessionToken: String
        let userId: String
        let encryptedKeys: String?
        let identityKey: String?
    }
    
    func login(username: String, authHash: String) async throws -> LoginResult {
        let body: [String: Any] = ["username": username, "auth_hash": authHash]
        let data = try await post("/auth/login", body: body)
        
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let token = json["session_token"] as? String,
              let userId = json["user_id"] as? String else {
            throw APIError.invalidResponse
        }
        sessionToken = token
        SecureStorage.shared.set(token, forKey: "session_token")
        UserDefaults.standard.removeObject(forKey: "session_token")
        refreshToken = json["refresh_token"] as? String
        if let refreshToken {
            SecureStorage.shared.set(refreshToken, forKey: "refresh_token")
            UserDefaults.standard.removeObject(forKey: "refresh_token")
        }
        
        return LoginResult(
            sessionToken: token,
            userId: userId,
            encryptedKeys: json["encrypted_keys"] as? String,
            identityKey: json["identity_key"] as? String
        )
    }
    
    struct RegisterResult {
        let sessionToken: String
        let userId: String
        let deviceId: String
    }

    func register(
        username: String,
        displayName: String,
        authHash: String,
        salt: String,
        identityKey: String,
        identityDHKey: String,
        identityPrivateEncrypted: String,
        signedPreKeyPublic: String,
        signedPreKeyPrivateEncrypted: String,
        signedPreKeySignature: String,
        oneTimePreKeys: [String]
    ) async throws -> RegisterResult {
        // Solve proof-of-work before registration
        let pow = try await solvePoW()

        var body: [String: Any] = [
            "username": username,
            "display_name": displayName,
            "auth_hash": authHash,
            "salt": salt,
            "identity_key": identityKey,
            "identity_dh_key": identityDHKey,
            "identity_private_encrypted": identityPrivateEncrypted,
            "signed_pre_key_public": signedPreKeyPublic,
            "signed_pre_key_private_encrypted": signedPreKeyPrivateEncrypted,
            "signed_pre_key_signature": signedPreKeySignature,
            "one_time_pre_keys": oneTimePreKeys,
        ]
        if let token = pow["token"], let nonce = pow["nonce"] {
            body["pow_token"] = token
            body["pow_nonce"] = nonce
        }
        let data = try await post("/auth/register", body: body)
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let token = json["session_token"] as? String,
              let userId = json["user_id"] as? String else {
            throw APIError.invalidResponse
        }
        sessionToken = token
        SecureStorage.shared.set(token, forKey: "session_token")
        UserDefaults.standard.removeObject(forKey: "session_token")
        refreshToken = json["refresh_token"] as? String
        if let refreshToken {
            SecureStorage.shared.set(refreshToken, forKey: "refresh_token")
            UserDefaults.standard.removeObject(forKey: "refresh_token")
        }
        return RegisterResult(
            sessionToken: token,
            userId: userId,
            deviceId: json["device_id"] as? String ?? ""
        )
    }
    
    func getConversations() async throws -> [[String: Any]] {
        let data = try await get("/messages/conversations")
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let conversations = json["conversations"] as? [[String: Any]] else {
            throw APIError.invalidResponse
        }
        return conversations
    }
    
    func getContacts() async throws -> [[String: Any]] {
        let data = try await get("/contacts")
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let contacts = json["contacts"] as? [[String: Any]] else {
            throw APIError.invalidResponse
        }
        return contacts
    }
    
    func getMe() async throws -> [String: Any] {
        let data = try await get("/me")
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw APIError.invalidResponse
        }
        return json
    }
    
    // MARK: - Public raw HTTP helpers (used by chat screens)
    
    func getRaw(_ path: String) async throws -> Data {
        return try await get(path)
    }
    
    func postRaw(_ path: String, body: [String: Any], method: String = "POST") async throws -> Data {
        var request = URLRequest(url: URL(string: "\(baseURL)\(path)")!)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token = sessionToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (data, response) = try await requestWithRetry(request)
        try validateResponse(response)
        return data
    }

    func uploadBinary(_ path: String, data: Data, headers: [String: String]) async throws -> Data {
        var request = URLRequest(url: URL(string: "\(baseURL)\(path)")!)
        request.httpMethod = "POST"
        if let token = sessionToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        for (key, value) in headers {
            request.setValue(value, forHTTPHeaderField: key)
        }
        request.httpBody = data
        let (respData, response) = try await requestWithRetry(request)
        try validateResponse(response)
        return respData
    }

    func deleteRaw(_ path: String) async throws -> Data {
        var request = URLRequest(url: URL(string: "\(baseURL)\(path)")!)
        request.httpMethod = "DELETE"
        if let token = sessionToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        let (data, response) = try await requestWithRetry(request)
        try validateResponse(response)
        return data
    }
    
    // MARK: - HTTP helpers
    
    private func get(_ path: String) async throws -> Data {
        var request = URLRequest(url: URL(string: "\(baseURL)\(path)")!)
        request.httpMethod = "GET"
        if let token = sessionToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        let (data, response) = try await requestWithRetry(request)
        try validateResponse(response)
        return data
    }
    
    private func post(_ path: String, body: [String: Any]) async throws -> Data {
        var request = URLRequest(url: URL(string: "\(baseURL)\(path)")!)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token = sessionToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (data, response) = try await requestWithRetry(request)
        try validateResponse(response)
        return data
    }
    
    private func validateResponse(_ response: URLResponse) throws {
        guard let http = response as? HTTPURLResponse else { throw APIError.invalidResponse }
        guard (200...299).contains(http.statusCode) else {
            throw APIError.httpError(http.statusCode)
        }
    }
    
    enum APIError: Error {
        case invalidResponse
        case httpError(Int)
    }

    /// Upload encrypted media to R2, returns the blob ID string
    func uploadMedia(_ encryptedData: Data) async throws -> String {
        let data = try await uploadBinary("/media/upload", data: encryptedData, headers: [
            "Content-Type": "application/octet-stream"
        ])
        if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let blobId = json["blob_id"] as? String {
            return blobId
        }
        // Fallback: some APIs return the id at the top level
        if let str = String(data: data, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines), !str.isEmpty {
            return str
        }
        throw APIError.invalidResponse
    }

    // MARK: - Proof-of-Work Solver

    func solvePoW() async throws -> [String: String] {
        let challengeData = try await getRaw("/features/pow/challenge")
        guard let json = try JSONSerialization.jsonObject(with: challengeData) as? [String: Any],
              let token = json["token"] as? String,
              let challenge = json["challenge"] as? String,
              let difficulty = json["difficulty"] as? Int else {
            return [:]
        }
        if difficulty == 0 { return ["token": token, "nonce": "0"] }

        for nonce in 0..<10_000_000 {
            let input = "\(challenge):\(nonce)"
            guard let inputData = input.data(using: .utf8) else { continue }
            var hash = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
            inputData.withUnsafeBytes { ptr in
                _ = CC_SHA256(ptr.baseAddress, CC_LONG(inputData.count), &hash)
            }
            if leadingZeroBits(hash) >= difficulty {
                return ["token": token, "nonce": String(nonce)]
            }
        }
        return [:]
    }

    private func leadingZeroBits(_ bytes: [UInt8]) -> Int {
        var bits = 0
        for b in bytes {
            if b == 0 { bits += 8; continue }
            var mask: UInt8 = 0x80
            while mask > 0 && (b & mask) == 0 {
                bits += 1
                mask >>= 1
            }
            return bits
        }
        return bits
    }

    // MARK: - Refresh Token Rotation

    func refreshSession() async throws -> Bool {
        let refresh = refreshToken ?? SecureStorage.shared.get(forKey: "refresh_token")
        guard let rt = refresh else { return false }
        var request = URLRequest(url: URL(string: "\(baseURL)/auth/refresh")!)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: ["refresh_token": rt])
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
            clearAuthState()
            return false
        }
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let newSession = json["session_token"] as? String else { return false }
        sessionToken = newSession
        refreshToken = json["refresh_token"] as? String
        SecureStorage.shared.set(newSession, forKey: "session_token")
        UserDefaults.standard.removeObject(forKey: "session_token")
        if let refreshToken {
            SecureStorage.shared.set(refreshToken, forKey: "refresh_token")
            UserDefaults.standard.removeObject(forKey: "refresh_token")
        } else {
            SecureStorage.shared.remove(forKey: "refresh_token")
        }
        return true
    }

    private func requestWithRetry(_ request: URLRequest) async throws -> (Data, URLResponse) {
        let (data, response) = try await session.data(for: request)
        if let http = response as? HTTPURLResponse, http.statusCode == 401 {
            // Try to refresh and retry once
            if try await refreshSession() {
                var retryReq = request
                if let token = sessionToken {
                    retryReq.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
                }
                return try await session.data(for: retryReq)
            }
            clearAuthState()
        }
        return (data, response)
    }

    private func clearAuthState() {
        sessionToken = nil
        refreshToken = nil
        SecureStorage.shared.remove(forKey: "session_token")
        SecureStorage.shared.remove(forKey: "refresh_token")
        UserDefaults.standard.removeObject(forKey: "session_token")
        UserDefaults.standard.removeObject(forKey: "refresh_token")
    }
}

File: android/app/src/main/kotlin/com/rocchat/network/APIClient.kt
package com.rocchat.network

import android.content.Context
import com.rocchat.crypto.SecureStorage
import org.json.JSONArray
import org.json.JSONObject
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.security.MessageDigest
import java.security.cert.X509Certificate
import javax.net.ssl.HttpsURLConnection
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

object APIClient {
    private const val BASE_URL = "https://chat.mocipher.com/api"

    var sessionToken: String? = null
    var refreshToken: String? = null
    private var appContext: Context? = null

    private val pinnedLeafFingerprints = mapOf(
        "chat.mocipher.com" to "9fe88e6203b5c859d5d5afad6b2efe52b3f01bb06ea1b140c43b1b77ebd89dbb",
        "rocchat-api.spoass.workers.dev" to "fced5d78a8da9d40a74f053b91aed908a2a0ef097a8878b002409be9b35eead1",
    )

    fun initialize(context: Context) {
        appContext = context.applicationContext
        val ctx = appContext ?: return
        refreshToken = SecureStorage.get(ctx, "refresh_token", "rocchat")
        if (refreshToken == null) {
            val legacy = ctx.getSharedPreferences("rocchat", Context.MODE_PRIVATE).getString("refresh_token", null)
            if (!legacy.isNullOrEmpty()) {
                refreshToken = legacy
                SecureStorage.set(ctx, "refresh_token", legacy)
                ctx.getSharedPreferences("rocchat", Context.MODE_PRIVATE).edit().remove("refresh_token").apply()
            }
        }
    }

    fun clearPersistedAuth() {
        val ctx = appContext ?: return
        SecureStorage.remove(ctx, "session_token")
        SecureStorage.remove(ctx, "refresh_token")
        ctx.getSharedPreferences("rocchat", Context.MODE_PRIVATE).edit()
            .remove("session_token")
            .remove("refresh_token")
            .apply()
    }

    private fun persistRefreshToken(token: String?) {
        val ctx = appContext ?: return
        if (token.isNullOrEmpty()) {
            SecureStorage.remove(ctx, "refresh_token")
            ctx.getSharedPreferences("rocchat", Context.MODE_PRIVATE).edit().remove("refresh_token").apply()
            return
        }
        SecureStorage.set(ctx, "refresh_token", token)
        ctx.getSharedPreferences("rocchat", Context.MODE_PRIVATE).edit().remove("refresh_token").apply()
    }

    private fun responseCodeWithPinning(conn: HttpURLConnection): Int {
        val code = conn.responseCode
        if (conn is HttpsURLConnection) {
            val host = conn.url.host.lowercase()
            val expected = pinnedLeafFingerprints[host]
            if (expected != null) {
                val leaf = conn.serverCertificates.firstOrNull() as? X509Certificate
                    ?: throw IOException("Pinned TLS certificate missing")
                val actual = MessageDigest.getInstance("SHA-256")
                    .digest(leaf.encoded)
                    .joinToString("") { "%02x".format(it) }
                if (!actual.equals(expected, ignoreCase = true)) {
                    throw IOException("Pinned TLS certificate mismatch")
                }
            }
        }
        return code
    }

    // ── Auth ──

    data class LoginResult(
        val sessionToken: String,
        val userId: String,
        val encryptedKeys: String?,
        val identityKey: String?,
    )

    suspend fun login(username: String, authHash: String): LoginResult {
        val body = JSONObject().apply {
            put("username", username)
            put("auth_hash", authHash)
        }
        val json = post("/auth/login", body)
        refreshToken = json.optString("refresh_token", null)
        persistRefreshToken(refreshToken)
        return LoginResult(
            sessionToken = json.getString("session_token"),
            userId = json.getString("user_id"),
            encryptedKeys = json.optString("encrypted_keys", null),
            identityKey = json.optString("identity_key", null),
        )
    }

    suspend fun register(
        username: String,
        displayName: String,
        authHash: String,
        salt: String,
        identityKey: String,
        identityDHKey: String = "",
        identityPrivateEncrypted: String,
        signedPreKeyPublic: String,
        signedPreKeyPrivateEncrypted: String,
        signedPreKeySignature: String,
        oneTimePreKeys: List<String>,
    ): JSONObject {
        // Solve proof-of-work before registration
        val pow = solvePoW()

        val body = JSONObject().apply {
            put("username", username)
            put("display_name", displayName)
            put("auth_hash", authHash)
            put("salt", salt)
            put("identity_key", identityKey)
            if (identityDHKey.isNotEmpty()) put("identity_dh_key", identityDHKey)
            put("identity_private_encrypted", identityPrivateEncrypted)
            put("signed_pre_key_public", signedPreKeyPublic)
            put("signed_pre_key_private_encrypted", signedPreKeyPrivateEncrypted)
            put("signed_pre_key_signature", signedPreKeySignature)
            put("one_time_pre_keys", JSONArray(oneTimePreKeys))
            pow["token"]?.let { put("pow_token", it) }
            pow["nonce"]?.let { put("pow_nonce", it) }
        }
        val json = post("/auth/register", body)
        refreshToken = json.optString("refresh_token", null)
        persistRefreshToken(refreshToken)
        return json
    }

    // ── Conversations ──

    data class ConversationMember(
        val userId: String,
        val username: String,
        val displayName: String,
        val avatarUrl: String?,
    )

    data class Conversation(
        val id: String,
        val type: String,
        val name: String?,
        val members: List<ConversationMember>,
        val lastMessageAt: String?,
        val lastMessageType: String?,
        val muted: Boolean,
        val archived: Boolean,
    )

    suspend fun getConversations(): List<Conversation> {
        val json = get("/messages/conversations")
        val arr = json.optJSONArray("conversations") ?: return emptyList()
        return (0 until arr.length()).mapNotNull { i ->
            val c = arr.getJSONObject(i)
            val archived = c.optBoolean("archived", false)
            if (archived) return@mapNotNull null
            val membersArr = c.optJSONArray("members") ?: JSONArray()
            Conversation(
                id = c.getString("id"),
                type = c.optString("type", "direct"),
                name = c.optString("name", null),
                members = (0 until membersArr.length()).map { j ->
                    val m = membersArr.getJSONObject(j)
                    ConversationMember(
                        userId = m.getString("user_id"),
                        username = m.optString("username", ""),
                        displayName = m.optString("display_name", ""),
                        avatarUrl = m.optString("avatar_url", null),
                    )
                },
                lastMessageAt = c.optString("last_message_at", null),
                lastMessageType = c.optString("last_message_type", null).ifEmpty { null },
                muted = c.optBoolean("muted", false),
                archived = archived,
            )
        }
    }

    data class ChatMessage(
        val id: String,
        val conversationId: String,
        val senderId: String,
        val ciphertext: String,
        val iv: String,
        val ratchetHeader: String,
        val messageType: String,
        val createdAt: String,
        val expiresAt: Long? = null,
        var status: String = "sent",
        val reactions: String? = null,
    )

    suspend fun getMessages(conversationId: String): List<ChatMessage> {
        val json = get("/messages/$conversationId")
        val arr = json.optJSONArray("messages") ?: return emptyList()
        return (0 until arr.length()).map { i ->
            val m = arr.getJSONObject(i)
            ChatMessage(
                id = m.getString("id"),
                conversationId = m.optString("conversation_id", conversationId),
                senderId = m.getString("sender_id"),
                ciphertext = m.optString("ciphertext", ""),
                iv = m.optString("iv", ""),
                ratchetHeader = m.optString("ratchet_header", ""),
                messageType = m.optString("message_type", "text"),
                createdAt = m.optString("created_at", ""),
                expiresAt = if (m.has("expires_at") && !m.isNull("expires_at")) m.getLong("expires_at") else null,
                status = m.optString("status", "sent"),
            )
        }
    }

    suspend fun sendMessage(
        conversationId: String,
        ciphertext: String,
        iv: String,
        ratchetHeader: String,
        messageType: String = "text",
        expiresIn: Int = 0,
        replyTo: String? = null,
    ): JSONObject {
        val body = JSONObject().apply {
            put("conversation_id", conversationId)
            put("ciphertext", ciphertext)
            put("iv", iv)
            put("ratchet_header", ratchetHeader)
            put("message_type", messageType)
            if (expiresIn > 0) put("expires_in", expiresIn)
            if (!replyTo.isNullOrEmpty()) put("reply_to", replyTo)
        }
        return post("/messages/send", body)
    }

    suspend fun createConversation(type: String, memberIds: List<String>, encryptedMeta: String? = null): String {
        val body = JSONObject().apply {
            put("type", type)
            put("member_ids", JSONArray(memberIds))
            if (encryptedMeta != null) put("encrypted_meta", encryptedMeta)
        }
        val json = post("/messages/conversations", body)
        return json.getString("conversation_id")
    }

    suspend fun deleteConversation(conversationId: String): JSONObject = delete("/messages/conversations/$conversationId")

    suspend fun muteConversation(conversationId: String): Boolean {
        val json = post("/messages/conversations/$conversationId/mute", JSONObject())
        return json.optBoolean("muted", false)
    }

    suspend fun archiveConversation(conversationId: String): Boolean {
        val json = post("/messages/conversations/$conversationId/archive", JSONObject())
        return json.optBoolean("archived", false)
    }

    // ── Contacts ──

    data class UserSearchResult(
        val userId: String,
        val username: String,
        val displayName: String,
    )

    suspend fun searchUsers(query: String): List<UserSearchResult> {
        val json = get("/contacts/search?q=$query")
        val arr = json.optJSONArray("results") ?: return emptyList()
        return (0 until arr.length()).map { i ->
            val u = arr.getJSONObject(i)
            UserSearchResult(
                userId = u.getString("userId"),
                username = u.getString("username"),
                displayName = u.getString("displayName"),
            )
        }
    }

    // ── QR Auth ──

    suspend fun authorizeQrToken(token: String): JSONObject {
        val body = JSONObject().apply { put("qr_token", token) }
        return post("/auth/qr/authorize", body)
    }

    // ── Profile ──

    suspend fun getMe(): JSONObject = get("/me")

    suspend fun updateSettings(settings: Map<String, Any>): JSONObject {
        return post("/me/settings", JSONObject(settings), method = "PATCH")
    }

    // ── HTTP Helpers ──

    suspend fun uploadAvatar(data: ByteArray, contentType: String = "image/jpeg"): JSONObject = withContext(Dispatchers.IO) {
        val conn = (URL("$BASE_URL/me/avatar").openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            sessionToken?.let { setRequestProperty("Authorization", "Bearer $it") }
            setRequestProperty("Content-Type", contentType)
            doOutput = true
            connectTimeout = 30_000
            readTimeout = 30_000
        }
        try {
            conn.outputStream.use { it.write(data) }
            val code = responseCodeWithPinning(conn)
            val body = (if (code in 200..299) conn.inputStream else conn.errorStream)
                ?.bufferedReader()?.use { it.readText() } ?: "{}"
            if (code !in 200..299) throw IOException("HTTP $code: $body")
            JSONObject(body)
        } finally {
            conn.disconnect()
        }
    }

    suspend fun uploadMedia(conversationId: String, data: ByteArray, filename: String, mimeType: String): String = withContext(Dispatchers.IO) {
        val conn = (URL("$BASE_URL/media/upload").openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            sessionToken?.let { setRequestProperty("Authorization", "Bearer $it") }
            setRequestProperty("Content-Type", "application/octet-stream")
            setRequestProperty("x-conversation-id", conversationId)
            setRequestProperty("x-encrypted-filename", filename)
            setRequestProperty("x-encrypted-mimetype", mimeType)
            doOutput = true
            connectTimeout = 30_000
            readTimeout = 30_000
        }
        try {
            conn.outputStream.use { it.write(data) }
            val code = responseCodeWithPinning(conn)
            val body = (if (code in 200..299) conn.inputStream else conn.errorStream)
                ?.bufferedReader()?.use { it.readText() } ?: "{}"
            if (code !in 200..299) throw IOException("HTTP $code: $body")
            JSONObject(body).getString("mediaId")
        } finally {
            conn.disconnect()
        }
    }

    suspend fun postPublic(path: String, body: JSONObject): JSONObject = post(path, body)

    suspend fun get(path: String): JSONObject = withContext(Dispatchers.IO) {
        executeWithRetry {
            val conn = (URL("$BASE_URL$path").openConnection() as HttpURLConnection).apply {
                requestMethod = "GET"
                sessionToken?.let { setRequestProperty("Authorization", "Bearer $it") }
                connectTimeout = 15_000
                readTimeout = 15_000
            }
            try {
                val code = responseCodeWithPinning(conn)
                val body = (if (code in 200..299) conn.inputStream else conn.errorStream)
                    ?.bufferedReader()?.use { it.readText() } ?: "{}"
                if (code !in 200..299) throw IOException("HTTP $code: $body")
                JSONObject(body)
            } finally {
                conn.disconnect()
            }
        }
    }

    suspend fun getArray(path: String): JSONArray = withContext(Dispatchers.IO) {
        val conn = (URL("$BASE_URL$path").openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            sessionToken?.let { setRequestProperty("Authorization", "Bearer $it") }
            connectTimeout = 15_000
            readTimeout = 15_000
        }
        try {
            val code = responseCodeWithPinning(conn)
            val body = (if (code in 200..299) conn.inputStream else conn.errorStream)
                ?.bufferedReader()?.use { it.readText() } ?: "[]"
            if (code !in 200..299) throw IOException("HTTP $code: $body")
            JSONArray(body)
        } finally {
            conn.disconnect()
        }
    }

    suspend fun getRawBytes(path: String): ByteArray = withContext(Dispatchers.IO) {
        val conn = (URL("$BASE_URL$path").openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            sessionToken?.let { setRequestProperty("Authorization", "Bearer $it") }
            connectTimeout = 15_000
            readTimeout = 30_000
        }
        try {
            val code = responseCodeWithPinning(conn)
            if (code !in 200..299) throw IOException("HTTP $code")
            conn.inputStream.readBytes()
        } finally {
            conn.disconnect()
        }
    }

    suspend fun delete(path: String): JSONObject = withContext(Dispatchers.IO) {
        val conn = (URL("$BASE_URL$path").openConnection() as HttpURLConnection).apply {
            requestMethod = "DELETE"
            sessionToken?.let { setRequestProperty("Authorization", "Bearer $it") }
            connectTimeout = 15_000
            readTimeout = 15_000
        }
        try {
            val code = responseCodeWithPinning(conn)
            val body = (if (code in 200..299) conn.inputStream else conn.errorStream)
                ?.bufferedReader()?.use { it.readText() } ?: "{}"
            if (code !in 200..299) throw IOException("HTTP $code: $body")
            JSONObject(body)
        } finally {
            conn.disconnect()
        }
    }

    suspend fun post(path: String, body: JSONObject, method: String = "POST"): JSONObject =
        withContext(Dispatchers.IO) {
            executeWithRetry {
                val conn = (URL("$BASE_URL$path").openConnection() as HttpURLConnection).apply {
                    requestMethod = if (method == "PATCH") "POST" else method
                    if (method == "PATCH") setRequestProperty("X-HTTP-Method-Override", "PATCH")
                    setRequestProperty("Content-Type", "application/json; charset=utf-8")
                    sessionToken?.let { setRequestProperty("Authorization", "Bearer $it") }
                    doOutput = true
                    connectTimeout = 15_000
                    readTimeout = 15_000
                }
                try {
                    conn.outputStream.bufferedWriter().use { it.write(body.toString()) }
                    val code = responseCodeWithPinning(conn)
                    val respBody = (if (code in 200..299) conn.inputStream else conn.errorStream)
                        ?.bufferedReader()?.use { it.readText() } ?: "{}"
                    if (code !in 200..299) throw IOException("HTTP $code: $respBody")
                    JSONObject(respBody)
                } finally {
                    conn.disconnect()
                }
            }
        }

    suspend fun uploadMedia(data: ByteArray): String = withContext(Dispatchers.IO) {
        val conn = (URL("$BASE_URL/media/upload").openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            setRequestProperty("Content-Type", "application/octet-stream")
            sessionToken?.let { setRequestProperty("Authorization", "Bearer $it") }
            doOutput = true
            connectTimeout = 30_000
            readTimeout = 30_000
        }
        try {
            conn.outputStream.use { it.write(data) }
            val code = responseCodeWithPinning(conn)
            val respBody = (if (code in 200..299) conn.inputStream else conn.errorStream)
                ?.bufferedReader()?.use { it.readText() } ?: "{}"
            if (code !in 200..299) throw IOException("HTTP $code: $respBody")
            val json = JSONObject(respBody)
            json.optString("blob_id", respBody.trim())
        } finally {
            conn.disconnect()
        }
    }

    // ── Proof-of-Work Solver ──

    private suspend fun solvePoW(): Map<String, String> = withContext(Dispatchers.IO) {
        try {
            val conn = (URL("$BASE_URL/features/pow/challenge").openConnection() as HttpURLConnection).apply {
                requestMethod = "GET"
                connectTimeout = 15_000
                readTimeout = 15_000
            }
            val json = try {
                val code = responseCodeWithPinning(conn)
                val body = (if (code in 200..299) conn.inputStream else conn.errorStream)
                    ?.bufferedReader()?.use { it.readText() } ?: "{}"
                if (code !in 200..299) return@withContext emptyMap()
                JSONObject(body)
            } finally {
                conn.disconnect()
            }
            val token = json.optString("token", "") ?: ""
            val challenge = json.optString("challenge", "") ?: ""
            val difficulty = json.optInt("difficulty", 0)
            if (token.isEmpty() || challenge.isEmpty()) return@withContext emptyMap()
            if (difficulty == 0) return@withContext mapOf("token" to token, "nonce" to "0")

            val md = MessageDigest.getInstance("SHA-256")
            for (nonce in 0 until 10_000_000) {
                md.reset()
                val digest = md.digest("$challenge:$nonce".toByteArray(Charsets.UTF_8))
                if (leadingZeroBits(digest) >= difficulty) {
                    return@withContext mapOf("token" to token, "nonce" to nonce.toString())
                }
            }
            emptyMap()
        } catch (_: Exception) {
            emptyMap()
        }
    }

    private fun leadingZeroBits(bytes: ByteArray): Int {
        var bits = 0
        for (b in bytes) {
            val unsigned = b.toInt() and 0xFF
            if (unsigned == 0) { bits += 8; continue }
            var mask = 0x80
            while (mask > 0 && (unsigned and mask) == 0) { bits++; mask = mask shr 1 }
            return bits
        }
        return bits
    }

    // ── Refresh Token Rotation ──

    private suspend fun refreshSession(): Boolean = withContext(Dispatchers.IO) {
        val rt = refreshToken ?: SecureStorage.get(appContext ?: return@withContext false, "refresh_token", "rocchat") ?: return@withContext false
        val conn = (URL("$BASE_URL/auth/refresh").openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            setRequestProperty("Content-Type", "application/json; charset=utf-8")
            doOutput = true
            connectTimeout = 15_000
            readTimeout = 15_000
        }
        try {
            conn.outputStream.bufferedWriter().use {
                it.write(JSONObject().apply { put("refresh_token", rt) }.toString())
            }
            val code = responseCodeWithPinning(conn)
            if (code !in 200..299) {
                clearPersistedAuth()
                return@withContext false
            }
            val body = conn.inputStream.bufferedReader().use { it.readText() }
            val json = JSONObject(body)
            sessionToken = json.optString("session_token", null)
            refreshToken = json.optString("refresh_token", null)
            persistRefreshToken(refreshToken)
            sessionToken != null
        } catch (_: Exception) {
            false
        } finally {
            conn.disconnect()
        }
    }

    private suspend fun <T> executeWithRetry(block: () -> T): T {
        return try {
            block()
        } catch (e: IOException) {
            if (e.message?.startsWith("HTTP 401") == true && refreshSession()) {
                block()
            } else {
                throw e
            }
        }
    }
}
